#This patch makes easier to track qmail delivery logs
 
#before patch: 
#Fri Jan 25 15:42:41 2013  starting delivery 309: msg 1319022 to remote islam@surgate.com
#Fri Jan 25 15:42:41 2013  status: local 0/400 remote 1/400
#Fri Jan 25 15:42:46 2013  delivery 309: deferral: Sorry,_I_wasn't_able_to_establish_an_SMTP_connection._(#4.4.1)/
 
#Wed Jan 23 10:50:29 2013  starting delivery id 17069: msgid 5967  to remote islam@surgatelabs.com
#Wed Jan 23 10:50:29 2013  delivery status: local 0/500 remote 1/500 subq0 0/50 subq1 0/100 subq2 0/150
#Wed Jan 23 10:50:29 2013  delivery id 17069: success: 195.214.145.107_accepted_message./Remote_host_said:_250_ok_1358930587_qp_14635/
 
#after applying this patch the log becomes more beatiful :)
#Thu Jan 24 12:47:07 2013  starting delivery 6: msg 917 to remote islam@surgate.com
#Thu Jan 24 12:47:07 2013  status: local 0/10 remote 1/20
#Thu Jan 24 12:47:08 2013  delivery 6: deferral: Sorry,_I_wasn't_able_to_establish_an_SMTP_connection._(#4.4.1)from:surgatelabs@surgate.com_to:islam@surgate.com_>_195.214.145.108
 
#Wed Jan 23 10:50:29 2013  starting delivery id 17070: msgid 5973 from islam@surgatelabs.com to remote islam@surgate.com
#Wed Jan 23 10:50:29 2013  delivery status: local 0/500 remote 1/500 subq0 0/50 subq1 0/100 subq2 0/150
#Wed Jan 23 10:50:29 2013  delivery id 17070: success: <from:islam@surgatelabs.com_to:islam@surgate.com>_195.214.145.107_accepted_message./Remote_host_said:_250_ok_1358930588_qp_14643/
 
--- qmail-remote.c.orig 2013-01-24 17:48:20.759853414 +0200
+++ qmail-remote.c  2013-01-24 13:45:42.979869328 +0200
@@ -44,6 +44,8 @@
 stralloc host = {0};
 stralloc sender = {0};
  
+char remoteIP [128] = ""; /* added by Surgate Labs  */
+
 saa reciplist = {0};
  
 struct ip_address partner;
@@ -56,16 +58,26 @@
 ch = sa->s[i]; if (ch < 33) ch = '?'; if (ch > 126) ch = '?';
 if (substdio_put(subfdoutsmall,&ch,1) == -1) _exit(0); } }
  
+void outhost()
+{  
+  char x[IPFMT];
+  if (substdio_put(subfdoutsmall,x,ip_fmt(x,&partner)) == -1) _exit(0);
+}
+
 void temp_nomem() { out("ZOut of memory. (#4.3.0)\n"); zerodie(); }
 void temp_oserr() { out("Z\
 System resources temporarily unavailable. (#4.3.0)\n"); zerodie(); }
-void temp_noconn() { out("Z\
-Sorry, I wasn't able to establish an SMTP connection. (#4.4.1)\n"); zerodie(); }
+void temp_noconn() {
+   out("ZSorry, I wasn't able to establish an SMTP connection. (#4.4.1)"); out("from:"); outsafe(&sender); out(" to:"); outsafe(&reciplist.sa[0]); out(" > ");  out(remoteIP); 
+   zerodie(); 
+}
 void temp_read() { out("ZUnable to read message. (#4.3.0)\n"); zerodie(); }
-void temp_dnscanon() { out("Z\
-CNAME lookup failed temporarily. (#4.4.3)\n"); zerodie(); }
-void temp_dns() { out("Z\
-Sorry, I couldn't find any host by that name. (#4.1.2)\n"); zerodie(); }
+void temp_dnscanon() {     
+   out("ZCNAME lookup failed temporarily. (#4.4.3)\n"); 
+   zerodie(); 
+}
+void temp_dns() {  out("ZSorry, I couldn't find any host by that name. (#4.1.2)\n"); out(" from:"); outsafe(&sender); out(" to:"); outsafe(&reciplist.sa[0]); out("> ");  outhost(); 
+ zerodie(); }
 void temp_chdir() { out("Z\
 Unable to switch to home directory. (#4.3.0)\n"); zerodie(); }
 void temp_control() { out("Z\
@@ -74,24 +86,17 @@
 SMTP cannot transfer messages with partial final lines. (#5.6.2)\n"); zerodie(); }
 void perm_usage() { out("D\
 I (qmail-remote) was invoked improperly. (#5.3.5)\n"); zerodie(); }
-void perm_dns() { out("D\
-Sorry, I couldn't find any host named ");
-outsafe(&host);
-out(". (#5.1.2)\n"); zerodie(); }
-void perm_nomx() { out("D\
-Sorry, I couldn't find a mail exchanger or IP address. (#5.4.4)\n");
+void perm_dns() { outsafe(&sender); out(" to:"); outsafe(&reciplist.sa[0]); out("> ");  outhost(); out(" DSorry, I couldn't find any host named  (#5.1.2)\n");
+ zerodie(); }
+void perm_nomx() { 
+outsafe(&reciplist.sa[0]); out("> ");  outhost(); out(" DSorry, I couldn't find a mail exchanger or IP address. (#5.4.4)\n");
 zerodie(); }
-void perm_ambigmx() { out("D\
-Sorry. Although I'm listed as a best-preference MX or A for that host,\n\
-it isn't in my control/locals file, so I don't treat it as local. (#5.4.6)\n");
+void perm_ambigmx() { 
+   out("DSorry. Although I'm listed as a best-preference MX or A for that host,\n\
+it isn't in my own database and domain route table or the mail was not relayed, so I don't treat it as local. (#5.4.6)\n");
+   out(" from:"); outsafe(&sender); out(" to:"); outsafe(&reciplist.sa[0]); out("> ");  outhost(); 
 zerodie(); }
  
-void outhost()
-{
-  char x[IPFMT];
-  if (substdio_put(subfdoutsmall,x,ip_fmt(x,&partner)) == -1) _exit(0);
-}
-
 int flagcritical = 0;
  
 void dropped() {
@@ -222,21 +227,33 @@
   int flagbother;
   int i;
   
-  if (smtpcode() != 220) quit("ZConnected to "," but greeting failed");
+  if (smtpcode() != 220) {
+   out("s<from:"); outsafe(&sender); out(" to:"); outsafe(&reciplist.sa[0]); out("> ");  outhost();
+   quit("ZConnected to "," but greeting failed");
+  }
   
   substdio_puts(&smtpto,"HELO ");
   substdio_put(&smtpto,helohost.s,helohost.len);
   substdio_puts(&smtpto,"\r\n");
   substdio_flush(&smtpto);
-  if (smtpcode() != 250) quit("ZConnected to "," but my name was rejected");
+  if (smtpcode() != 250) { 
+   out("s<from:"); outsafe(&sender); out(" to:"); outsafe(&reciplist.sa[0]); out("> ");  outhost();
+   quit("ZConnected to "," but my name was rejected");
+  }
   
   substdio_puts(&smtpto,"MAIL FROM:<");
   substdio_put(&smtpto,sender.s,sender.len);
   substdio_puts(&smtpto,">\r\n");
   substdio_flush(&smtpto);
   code = smtpcode();
-  if (code >= 500) quit("DConnected to "," but sender was rejected");
-  if (code >= 400) quit("ZConnected to "," but sender was rejected");
+  if (code >= 500) {
+   out("h<from:"); outsafe(&sender); out(" to:"); outsafe(&reciplist.sa[0]); out("> ");  outhost();
+   quit("DConnected to "," but sender was rejected");
+  }
+  if (code >= 400) {
+       out("s<from:"); outsafe(&sender); out(" to:"); outsafe(&reciplist.sa[0]); out("> ");  outhost();
+   quit("ZConnected to "," but sender was rejected");
+  }
   
   flagbother = 0;
   for (i = 0;i < reciplist.len;++i) {
@@ -246,15 +263,15 @@
     substdio_flush(&smtpto);
     code = smtpcode();
     if (code >= 500) {
-      out("h"); outhost(); out(" does not like recipient.\n");
-      outsmtptext(); zero();
+   out("h<from:"); outsafe(&sender); out(" to:"); outsafe(&reciplist.sa[i]); out("> ");  outhost(); out(" does not like recipient\n");
+       outsmtptext(); zero();
     }
     else if (code >= 400) {
-      out("s"); outhost(); out(" does not like recipient.\n");
-      outsmtptext(); zero();
+   out("s<from:"); outsafe(&sender); out(" to:"); outsafe(&reciplist.sa[0]); out("> ");  outhost(); out(" does not like recipient\n");
+   outsmtptext(); zero();
     }
     else {
-      out("r"); zero();
+      out("r<from:"); outsafe(&sender); out(" to:"); outsafe(&reciplist.sa[0]); out("> "); zero();
       flagbother = 1;
     }
   }
@@ -262,14 +279,26 @@
   
   substdio_putsflush(&smtpto,"DATA\r\n");
   code = smtpcode();
-  if (code >= 500) quit("D"," failed on DATA command");
-  if (code >= 400) quit("Z"," failed on DATA command");
+  if (code >= 500) {
+       out("D<from:"); outsafe(&sender); out(" to:"); outsafe(&reciplist.sa[0]); out("> ");
+   quit("D"," failed on DATA command");
+  }
+  if (code >= 400) {
+       out("Z<from:"); outsafe(&sender); out(" to:"); outsafe(&reciplist.sa[0]); out("> ");
+   quit("Z"," failed on DATA command");
+  }
   
   blast();
   code = smtpcode();
   flagcritical = 0;
-  if (code >= 500) quit("D"," failed after I sent the message");
-  if (code >= 400) quit("Z"," failed after I sent the message");
+  if (code >= 500) {
+       out("D<from:"); outsafe(&sender); out(" to:"); outsafe(&reciplist.sa[0]); out("> "); 
+   quit("D"," failed after I sent the message");
+  }
+  if (code >= 400) {
+       out("Z<from:"); outsafe(&sender); out(" to:"); outsafe(&reciplist.sa[0]); out("> "); 
+       quit("Z"," failed after I sent the message");
+   }
   quit("K"," accepted message");
 }
  
@@ -343,7 +372,7 @@
   if (argc < 4) perm_usage();
   if (chdir(auto_qmail) == -1) temp_chdir();
   getcontrols();
- 
+  
   
   if (!stralloc_copys(&host,argv[1])) temp_nomem();
   
@@ -390,6 +419,8 @@
       if (ip.len <= 0) temp_dns();
   }
   
+  ip_fmt(remoteIP, &ip.ix->ip); /*surgate labs islam */
+   
   if (ip.len <= 0) perm_nomx();
   
   prefme = 100000;
